### ISSUE TITLE
MRD vs MS-STFTD

### ISSUE THREAD

**Issue by Yagelmx** [jishengpeng/WavTokenizer#10]
Hello! I am in awe at the results you achieved, and I am trying to understand the discriminator setup chosen in your work. I am familiar with the MSD and MPD as presented in HiFiGAN, but i'm having a hard time differentiating between the STFT/spectrogram discriminators.

In the paper, you specify that MRD and MS-STFTD are employed. Indeed in your code, I see [Encodec's discriminator](https://github.com/jishengpeng/WavTokenizer/blob/main/encoder/msstftd.py) and [DAC's MRD](https://github.com/jishengpeng/WavTokenizer/blob/2b37313674006f93ab528bf1d8f168b093f8dfb6/decoder/discriminators.py#L200) but changed to only be on magnitude spectrograms instead of complex ones as in the [original](https://github.com/jishengpeng/WavTokenizer/blob/2b37313674006f93ab528bf1d8f168b093f8dfb6/decoder/discriminator_dac.py#L167)? Also, I think the original MRD has band splitting which is absent in the new implementation? correct me if I'm mistaken, but why would we do the `abs` and omit bands? And besides the discrepancy, could you please elaborate what exactly is the key difference between the two? it seems both take spectrograms at multiple scales/resolutions.

**Comment by CookiePPP** [comment]
You can see in [here](https://github.com/jishengpeng/WavTokenizer/blob/2b37313674006f93ab528bf1d8f168b093f8dfb6/decoder/experiment.py#L10-L12) the discriminators used are;

1. a normal MultiPeriodDiscriminator from `.decoder.discriminators.MultiPeriodDiscriminator`
2. a normal single-band-amplitude MultiResolutionDiscriminator from `.decoder.discriminators.DiscriminatorR`
3. a multi-band-complex MultiResolutionDiscriminator from `.decoder.discriminator_dac.MRD`
4. another normal MultiPeriodDiscriminator from `.decoder.discriminator_dac.MPD` which appears to be a duplicate of the first one?

---

`.encoder.msstftd.MultiScaleSTFTDiscriminator` is not imported or used anywhere in this repo.

`.decoder.discriminator_dac.MSD` MultiScaleDiscriminator is not used since `.decoder.discriminator_dac.DACDiscriminator().rates` list is empty in all configs, so none of that type of discriminator are initialized.

---

PS: I have no relation to this repo and didn't contribute to it in any way. I'm just reading it at the same time as you.

**Comment by jishengpeng** [comment]
> You can see in [here](https://github.com/jishengpeng/WavTokenizer/blob/2b37313674006f93ab528bf1d8f168b093f8dfb6/decoder/experiment.py#L10-L12) the discriminators used for training WavTokenizer are;
> 
> 1. a normal MultiPeriodDiscriminator from [`.decoder.discriminators`](https://github.com/jishengpeng/WavTokenizer/blob/2b37313674006f93ab528bf1d8f168b093f8dfb6/decoder/discriminators.py#L9)
> 2. a normal single-band-amplitude-only MultiResolutionDiscriminator from [`.decoder.discriminators`](https://github.com/jishengpeng/WavTokenizer/blob/2b37313674006f93ab528bf1d8f168b093f8dfb6/decoder/discriminators.py#L101)
> 3. a multi-band-complex MRD from [`.decoder.discriminator_dac`](https://github.com/jishengpeng/WavTokenizer/blob/2b37313674006f93ab528bf1d8f168b093f8dfb6/decoder/discriminator_dac.py#L110)
> 4. another normal MPD from [`.decoder.discriminator_dac`](https://github.com/jishengpeng/WavTokenizer/blob/2b37313674006f93ab528bf1d8f168b093f8dfb6/decoder/discriminator_dac.py#L36) which appears to be a duplicate of the MultiPeriodDiscriminator at the top of this list.
> 
> MultiScaleSTFTDiscriminator from `.encoder.msstftd` is not imported or used anywhere in this repo.
> 
> the MultiScaleDiscriminator from `.decoder.discriminator_dac` is not used since `.decoder.discriminator_dac.DACDiscriminator().rates` list is empty in all configs, so none of that type of discriminator are initialized. Also this MSD class has it's [resampling layers disabled](https://github.com/jishengpeng/WavTokenizer/blob/2b37313674006f93ab528bf1d8f168b093f8dfb6/decoder/discriminator_dac.py#L93) so it doesn't work correctly.
> 
> PS: I have no relation to this repo and didn't contribute to it in any way. I'm just reading it at the same time as you.

There are minor bugs present; our intention was to replicate the discriminator from the DAC, but we inadvertently copied a different one. While the majority of @CookiePPP's points are accurate, it's important to note that another MPD operates on multiple scales and returns feature maps instead of x. Thus, it differs from the first MPD. Additionally, the ablation experiments mentioned in the paper are correct. Modifying lines 185 and 186 in the [experiments](https://github.com/jishengpeng/WavTokenizer/blob/main/decoder/experiment.py) file will aid in reconstruction.

**Comment by CookiePPP** [comment]
> Modifying lines 185 and 186 in the [experiments](https://github.com/jishengpeng/WavTokenizer/blob/main/decoder/experiment.py) file will aid in reconstruction.

https://github.com/jishengpeng/WavTokenizer/blob/e435b4002889b12a059ccfedca8ffbe76257681d/decoder/experiment.py#L87-L90

I don't think the dac discriminator is receiving gradients by the way, not sure if you're using pretrained weights or this is an oversight but the parameters aren't being given to the optimizer.

**Comment by jishengpeng** [comment]
> > Modifying lines 185 and 186 in the [experiments](https://github.com/jishengpeng/WavTokenizer/blob/main/decoder/experiment.py) file will aid in reconstruction.
> 
> https://github.com/jishengpeng/WavTokenizer/blob/e435b4002889b12a059ccfedca8ffbe76257681d/decoder/experiment.py#L87-L90
> 
> I don't think the dac discriminator weights are being updated by the way, not sure if you're using pretrained weights or this is another 'minor bug' but the parameters aren't being given to the optimizer.
> 
> (`self.dac.parameters()` is not being given to the optimizer however I think the param.requires_grad is still set to true for those parameters, so it's not clear to me if this is intentional.)

It is noteworthy that multiple versions of the wavtokenizer.

**Comment by WhiteTeaDragon** [comment]
@jishengpeng just to be clear: in Table 7 in the paper you report UTMOS with and without MSTFTD -- which of the four discriminators mentioned above was removed? Was it multi-band-complex MRD from [.decoder.discriminator_dac](https://github.com/jishengpeng/WavTokenizer/blob/2b37313674006f93ab528bf1d8f168b093f8dfb6/decoder/discriminator_dac.py#L110)?
